examples/control-vector-generator/control-vector-generator.cpp

```diff
@@ -0,0 +1,549 @@
+        std::vector<float> b_prev_tensor = b_tensor;
+
+        // matrix multiplication and renormalize
+        b_tensor = mul_mat(matrix, b_tensor, cb_data.n_embd);
+        normalize_inplace(b_tensor);
+
+        // convergence check
+        float diff = 0.0;
+        for (int i = 0; i < cb_data.n_embd; ++i) {
+            diff += std::pow(b_tensor[i] - b_prev_tensor[i], 2);
+        }
+        if (std::sqrt(diff) < tolerance) {
+            break;
+        }
+    }
+
+    return b_tensor;
+}
+
+// TODO translate to ggml
+static void pca(callback_data & cb_data) {
+    size_t n_threads = 8;
+    int    n_layers  = cb_data.v_diff.size();
+    std::vector<std::thread> threads;
+    cb_data.v_final.reserve(n_layers);
+    auto worker_function = [&](int worker_id) {
+        for (int il = worker_id; il < n_layers; il += n_threads) {
+            float * matrix = square_diff(cb_data, il);
+            std::vector<float> eigenvector = power_iteration(cb_data, matrix);
+            cb_data.v_final[il] = &eigenvector[0];
+            delete[] matrix;
+            printf("Done with layer %d\n", il);
+        }
+    };
+    for (int i = 0; i < n_threads; ++i) {
+        threads.emplace_back(worker_function, i);
+    }
+    for (auto & th : threads) th.join();
+    printf("Done with PCA.");
+}
+
+template <typename T>
+static std::string to_string(const T & val) {
+    std::stringstream ss;
+    ss << val;
+    return ss.str();
+}
+
+static void export_gguf(std::vector<float *> v_final, int n_embd, const std::string fname, const std::string model_hint) {
+    struct gguf_context * ctx = gguf_init_empty();
+
+    const std::string arch = "controlvector";
+    gguf_set_val_str(ctx, "general.architecture", arch.c_str());
+    gguf_set_val_str(ctx, (arch + ".model_hint").c_str(), model_hint.c_str());
+    gguf_set_val_i32(ctx, (arch + ".layer_count").c_str(), v_final.size());
+
+    // TODO customize mem size - I have no idea what this is supposed to be
+    struct ggml_init_params params = {
+        /*.mem_size   =*/ (ggml_tensor_overhead() * v_final.size())
+                            + (n_embd * v_final.size() * sizeof(float)),
+        /*.mem_buffer =*/ NULL,
+        /*.no_alloc   =*/ false,
+    };
+
+    struct ggml_context * ctx_data = ggml_init(params);
+
+    for (size_t i = 0; i < v_final.size(); ++i) {
+        // TODO this number is probably not right - figure out which layer is which
+        // the python implementation uses a dict to handle this, we don't know if it's 1, 2, 3, 4... or other
+        const std::string name = "direction." + to_string(i+1);
+
+        struct ggml_tensor * cur = ggml_new_tensor_1d(ctx_data, GGML_TYPE_F32, n_embd);
+
+        ggml_set_name(cur, name.c_str());
+
+        // TODO figure out how to set data - it's whining about buf != NULL when using the below commented line
+        //ggml_backend_tensor_set(cur, cb_data.v_final[i], 0, cb_data.n_embd * sizeof(float));
```

Comment:
If the context is for writing our tensors into gguf, you don't need to worry about `ggml_backend_tensor_set`.

`ggml_backend_tensor_set` is used when you want the tensor to be offloaded onto device (a GPU for example). When writing to gguf, tensor data must be on RAM. Using `memcpy` (or a for loop as you did) is OK.